test/integration/replicaset/replicaset_test.go

```diff
@@ -515,3 +519,682 @@ func TestRSSelectorImmutability(t *testing.T) {
+		t.Fatal(err)
+	}
+}
+
+func TestReadyAndAvailableReplicas(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-ready-and-available-replicas", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 3)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	// Make .spec.minReadySeconds long enough so that we can test the replicas are unavailable initially
+	rs.Spec.MinReadySeconds = 3600
+
+	// Create 3 pods
+	var podList []*v1.Pod
+	moment := metav1.Now()
+	for i := 0; i < 3; i++ {
+		// When creating the pod, manually set pod status as there is no kubelet in integration test env
+		pod := newPod(fmt.Sprintf("pod-%d", i+1), rs, v1.PodRunning, &moment, true)
+		pod.Labels = labelMap
+		pod.Spec.Containers = []v1.Container{
+			{
+				Name:  "nginx",
+				Image: "gcr.io/google_containers/nginx-slim:0.8",
+			},
+		}
+		podList = append(podList, pod)
+	}
+	createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, podList, ns.Name)
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	markAllPodsReady(t, c, rs, ns.Name)
+	waitAllPodsReady(t, c, rs, ns.Name)
+	waitRSStable(t, c, rs, ns.Name)
+
+	rs, err := c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(rs.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get replicaset %s: %v", rs.Name, err)
+	}
+	if rs.Status.Replicas != 3 {
+		t.Fatalf("Unexpected .Status.Replicas: Expected 3, saw %d\n", rs.Status.Replicas)
+	}
+	if rs.Status.ReadyReplicas != 3 {
+		t.Fatalf("Unexpected .Status.ReadyReplicas: Expected 3, saw %d\n", rs.Status.ReadyReplicas)
+	}
+	//Verify all pods are not available manually
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+
+	// Make .spec.minReadySeconds = 0 so that we can test the replicas are available
+	rs.Spec.MinReadySeconds = 0
+	rs, err = c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(rs.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get replicaset %s: %v", rs.Name, err)
+	}
+	rs, err = c.ExtensionsV1beta1().ReplicaSets(ns.Name).Update(rs)
+	if err != nil {
+		t.Fatalf("failed to update replicaset %s: %v", rs.Name, err)
+	}
+	waitRSStable(t, c, rs, ns.Name)
+
+	// After updating the RS and waiting RS to be stable, we still have to GET the RS
+	// for the RS to reflect newest .Status
+	rs, err = c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(rs.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get replicaset %s: %v", rs.Name, err)
+	}
+	if rs.Status.Replicas != 3 {
+		t.Fatalf("Unexpected .Status.Replicas: Expected 3, saw %d\n", rs.Status.Replicas)
+	}
+	if rs.Status.ReadyReplicas != 3 {
+		t.Fatalf("Unexpected .Status.ReadyReplicas: Expected 3, saw %d\n", rs.Status.ReadyReplicas)
+	}
+	// Verify all pods are available manually
```

Comment:
Whether a Pod is Available or not is something each observer decides on their own. It's not stored in the Pod anywhere. What we want to test here is that the RS will observe the Pods as Available, which we do by checking the RS Status.AvailableReplicas. Directly observing the Pods ourselves is not necessary; it doesn't say anything about the thing we want to test.

Reply:
Testing `Status.AvailableReplicas` should suffice. I am not able to get the test passed using `Status.AvailableReplicas` (the attribute will always be zero). I will spend more time debugging the attribute.

Reply:
I think `AvailableReplicas` is staying 0 because above when you set `MinReadySeconds` your change never takes effect because you overwrite the whole rs object with a Get. This will be fixed when you make all edits use `RetryOnConflict{Get;edit;Update}` style.

Reply:
Sound good! I will use `PollImmediate`.